File: app/src/androidTest/java/com/bbtest/utils/ShellCommand.java
package com.bbtest.utils;

import android.app.Instrumentation;
import android.os.ParcelFileDescriptor;

import androidx.test.InstrumentationRegistry;
import androidx.test.uiautomator.UiDevice;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.FileInputStream;
import java.io.InputStreamReader;

public class ShellCommand {

    public static String execCmd(String cmd) {
        String result = "";
        try {
            Process pro = Runtime.getRuntime().exec(cmd);
            BufferedReader bfr = new BufferedReader(new InputStreamReader(pro.getInputStream()));
            String line = "";
            while ((line = bfr.readLine()) != null) {
                result += line + "\n";
            }
            bfr.close();
        } catch (Throwable t) {
            t.printStackTrace();
        }
        return result;
    }

    public static String execCmd(String[] cmds) {
        String result = "";
        try {
            // 重定向错误流
            ProcessBuilder proBuilder = new ProcessBuilder(cmds);
            proBuilder.redirectErrorStream(true);
            Process pro = proBuilder.start();
            BufferedReader bfr = new BufferedReader(new InputStreamReader(pro.getInputStream()));
            String line = "";
            while ((line = bfr.readLine()) != null) {
                result += line + "\n";
            }
            bfr.close();
        } catch (Throwable t) {
            t.printStackTrace();
        }
        return result;
    }

    public static String execCmdByUiDevice(UiDevice device, String cmd) {
        String result = "";
        try {
            result = device.executeShellCommand(cmd);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return result;
    }

    public static String execCmdByNoRoot(String cmd) {
        String result = "";
        try {
            Instrumentation instrumentation = InstrumentationRegistry.getInstrumentation();
            ParcelFileDescriptor pfd = instrumentation.getUiAutomation().executeShellCommand(cmd);
            FileInputStream fis = new ParcelFileDescriptor.AutoCloseInputStream(pfd);
            BufferedReader bfr = new BufferedReader(new InputStreamReader(fis));
            String line = "";
            while ((line = bfr.readLine()) != null) {
                result += line + "\n";
            }
            fis.close();
            bfr.close();
        } catch (Throwable t) {
            t.printStackTrace();
        }
        return result;
    }

    public static String execCmdBySh(String cmd) {
        String result = "";
        try {
            // /system/bin/sh
            Process pro = Runtime.getRuntime().exec("sh");
            DataOutputStream dos = new DataOutputStream(pro.getOutputStream());
            dos.writeBytes(cmd + "\n");
            dos.flush();
            dos.close();
            BufferedReader bfr = new BufferedReader(new InputStreamReader(pro.getInputStream()));
            String line = "";
            while ((line = bfr.readLine()) != null) {
                if (!line.equals("")) {
                    result += line + "\n";
                }
            }
            pro.waitFor();
            pro.destroy();
            bfr.close();
        } catch (Throwable t) {
            t.printStackTrace();
        }
        return result;
    }

    public static String execCmdBySu(String cmd) {
        String result = "";
        try {
            Process pro = Runtime.getRuntime().exec("su");
            DataOutputStream dos = new DataOutputStream(pro.getOutputStream());
            dos.writeBytes(cmd + "\n");
            dos.flush();
            dos.writeBytes("exit\n");
            dos.flush();
//            dos.close();
            BufferedReader bfr = new BufferedReader(new InputStreamReader(pro.getInputStream()));
            String line = "";
            while ((line = bfr.readLine()) != null) {
                if (!line.equals("")) {
                    result += line + "\n";
                }
            }
            pro.waitFor();
            pro.destroy();
            bfr.close();
        } catch (Throwable t) {
            t.printStackTrace();
        }
        return result;
    }

}

File: app/src/androidTest/kotlin/com/bbtest/utils/ShellCommand.kt
package com.bbtest.utils

import android.app.Instrumentation
import android.os.ParcelFileDescriptor
import androidx.test.InstrumentationRegistry
import androidx.test.uiautomator.UiDevice
import java.io.BufferedReader
import java.io.DataOutputStream
import java.io.FileInputStream
import java.io.InputStreamReader

object ShellCommand {
    private inline fun readProcessOutput(block: () -> java.io.InputStream): String {
        val result = StringBuilder()
        try {
            BufferedReader(InputStreamReader(block())).use { reader ->
                reader.lineSequence().forEach { line ->
                    result.append(line).append('\n')
                }
            }
        } catch (t: Throwable) {
            t.printStackTrace()
        }
        return result.toString()
    }

    @JvmStatic
    fun execCmd(cmd: String): String = readProcessOutput {
        Runtime.getRuntime().exec(cmd).inputStream
    }

    @JvmStatic
    fun execCmd(cmds: Array<String>): String = readProcessOutput {
        ProcessBuilder(*cmds)
            .redirectErrorStream(true)
            .start()
            .inputStream
    }

    @JvmStatic
    fun execCmdByUiDevice(device: UiDevice, cmd: String): String {
        return try {
            device.executeShellCommand(cmd)
        } catch (e: Exception) {
            e.printStackTrace()
            ""
        }
    }

    @JvmStatic
    fun execCmdByNoRoot(cmd: String): String = readProcessOutput {
        val instrumentation: Instrumentation = InstrumentationRegistry.getInstrumentation()
        val parcelFileDescriptor: ParcelFileDescriptor = instrumentation.uiAutomation.executeShellCommand(cmd)
        ParcelFileDescriptor.AutoCloseInputStream(parcelFileDescriptor)
    }

    @JvmStatic
    fun execCmdBySh(cmd: String): String {
        val result = StringBuilder()
        try {
            val process = Runtime.getRuntime().exec("sh")
            DataOutputStream(process.outputStream).use { outputStream ->
                outputStream.writeBytes("$cmd\n")
                outputStream.flush()
            }
            BufferedReader(InputStreamReader(process.inputStream)).use { reader ->
                reader.lineSequence()
                    .filter { it.isNotEmpty() }
                    .forEach { line -> result.append(line).append('\n') }
            }
            process.waitFor()
            process.destroy()
        } catch (t: Throwable) {
            t.printStackTrace()
        }
        return result.toString()
    }

    @JvmStatic
    fun execCmdBySu(cmd: String): String {
        val result = StringBuilder()
        try {
            val process = Runtime.getRuntime().exec("su")
            DataOutputStream(process.outputStream).use { outputStream ->
                outputStream.writeBytes("$cmd\n")
                outputStream.flush()
                outputStream.writeBytes("exit\n")
                outputStream.flush()
            }
            BufferedReader(InputStreamReader(process.inputStream)).use { reader ->
                reader.lineSequence()
                    .filter { it.isNotEmpty() }
                    .forEach { line -> result.append(line).append('\n') }
            }
            process.waitFor()
            process.destroy()
        } catch (t: Throwable) {
            t.printStackTrace()
        }
        return result.toString()
    }
}
